_posts/2021-11-08-android-state-management-mvi.md

```diff
@@ -0,0 +1,423 @@
+
+This is how we would typically observe these `LiveData` objects for changes inside the Fragment:
+
+```kotlin
+class ProjectFragment: Fragment() {
+
+    fun observeViewModel() {
+        viewModel.isLoading.observe(this, {
+                // show loading or hide it
+        })
+        viewModel.project.observe(this, {
+                // show project
+        })
+        viewModel.error.observe(this, {
+                // show error or hide it
+        })
+    }
+}
+```
+
+From this example, it is difficult to know what the UI of the screen will look like at any point because each `LiveData`  observable can emit a new state at **any point in time**. If we start adding new functions to our `ViewModel` that emit new loading or error states, would you be able to describe what the UI will look like at a single point of time? 
+
+This can result in a [race condition](https://en.wikipedia.org/wiki/Race_condition), since these separate observables can emit state changes independently. What would the state look like if an error is emitted but there is currently a project loaded?
+
+Typically these kinds of issues manifest on the UI with overlapping states like in this example we had from the app:
+
+{:refdef: style="text-align: center;"}
+![I’m sure we’ve all seen or made a screen like this before 😬]({{site.baseurl}}/assets/images/android-state-management-mvi/mvvm_issues_overlapping.png){: width="250" }
+{: refdef}
+
+
+There are several drawbacks when using this approach of MVVM with separate LiveData objects:
+
+- It is hard to know at a single point in time, what the UI should look like, if there are multiple LiveData objects that change the state of the screen.
+- There is no single snapshot to be able to recreate the UI from easily.
+- This also then begs the question, are we handling all the potential cases here?
+
+One way to potentially improve on this behaviour, is to use data classes that contain the state information in one class, and expose only a singular LiveData object to the UI. This would help solve the issue of having multiple observables emitting different state that we’d need to keep track of, and its a building block of how MVI can work, as we will explore next. 
+
+# MVI / Unidirectional Data Flow
+
+Now that we understand where potential issues can come from with MVVM and multiple LiveData observables, let’s talk a bit about Unidirectional Data Flow or MVI (Model-View-Intent) and our experience with it.
+
+With Unidirectional Data Flow everything is modelled around a single state object. This state is immutable, to prevent unwanted side effects. There is a clear flow of how data changes through the system, since it only happens in a single direction.
+
+
+{:refdef: style="text-align: center;"}
+<figure>
+  <img src="{{{site.baseurl}}/assets/images/android-state-management-mvi/unidirectional_data_flow.png" alt="my alt text"/>
```

Comment:
```suggestion
  <img src="{{site.baseurl}}/assets/images/android-state-management-mvi/unidirectional_data_flow.png" alt="my alt text"/>
```
Jekyll was throwing a Liquid syntax error on this. Also, any chance this alt text can be updated?

Reply:
```suggestion
  <img src="{{site.baseurl}}/assets/images/android-state-management-mvi/unidirectional_data_flow.png" alt="Diagram showing Unidirectional Data Flow"/>
```